### ISSUE TITLE
RxJava support for responses

### ISSUE THREAD

**Issue by jpiche** [koush/ion#545]
Being able to return an Observable—from [RxJava](https://github.com/ReactiveX/RxJava)—for a response instead of dealing with callbacks could save a lot of boilerplate code. Instead of the existing `setCallback` method, like the following,

```java
Ion.with(context)
  .load("POST", url)
  .setJsonObjectBody(json)
  .asJsonObject()
  .withResponse()
  .setCallback(callback);
```

I think something like `observable()` might work well. Like so:

```java

Ion.with(context)
  .load("POST", url)
  .setJsonObjectBody(json)
  .asJsonObject()
  .withResponse()
  .observable()
  .subscribe(
    response -> { /* code */ },
    error -> { /* code */ }
  );
```

**Comment by schwiz** [comment]
Duplicate request with better API suggestion. https://github.com/koush/ion/issues/395

**Comment by writtmeyer** [comment]
Yes, that indeed would be a nice addition. Nut sure, though, if this wouldn't actually belong to the AndroidAsync project (https://github.com/koush/AndroidAsync), since the Future class is in there.

**Comment by koush** [comment]
You could write your own Observable that implements FutureCallback, no?

.then(new ObservableThing())
.subscribe(...)

.then is the same as .setCallback, but it returns the value that you passed in, for handy chaining.

**Comment by schwiz** [comment]
Not quite following you, do you have a full example. Would the onSubscribe of the ObservableThingThatImplementsFutureCallback actually kick off the ion request or would it be a no-op?

**Comment by koush** [comment]
Oh, it would be a noop.

**Comment by jpiche** [comment]
Thanks for the idea @koush. I'll try doing something like that.

**Comment by jpiche** [comment]
I was probably a bit hasty in opening this issue. Since Ion returns a `Future`, `Observable.from(future)` works great.

**Comment by schwiz** [comment]
@jpiche are you able to use that method and still remain async? I'm having trouble figuring out how to get the result from the future without blocking using the way you suggest.

**Comment by jpiche** [comment]
@schwiz, try applying a scheduler to the observable. What I'm doing is something like

```java
Observable.from(ionFuture)
  .map(response -> { /* some stuff */ })
  .subscribeOn(Schedulers.io())
  .observeOn(AndroidSchedulers.mainThread())
  .subscribe(...);
```

**Comment by koush** [comment]
I'd not recommend using Observable.from(future) since that will block a thread somewhere calling future.get.

If it's possible to implement Observable that waits for a callback, that would be ideal. I'm thinking about ways to implement this, as it's been a common request.

My main concern is implementing it, without requiring rxjava as a new dependency to all ion users.

**Comment by schwiz** [comment]
So far the cleanest thing for me is to have the ion callback inside the
onSubscibe of an Observable and then call subscriber.onNext when you get
the callback.

**Comment by schwiz** [comment]
You know you could at least keep rxjava from being a strict dependency with gradle builds pretty easily with a little refactoring you could even make gson optional the same way.

**Comment by Nilzor** [comment]
>I'd not recommend using Observable.from(future) since that will block a thread somewhere calling future.get.

I'm not sure this is right, or I misunderstand you. `Observable.from(future)` does not block - you're just converting one async method to another. `Observable.from(future).toBlocking()` would on the other hand block.

The problem, as I've come to understand, is not with `.from()` but with the fact that Ion is _hot_ while Retrofit is _cold_ ([hot vs cold])(http://www.introtorx.com/content/v1.0.10621.0/14_HotAndColdObservables.html)). Well - it's only a problem if you care about interchangability.  I'm actually inclined to change my issue #395 to address that matter instead - changin the title to something like "There should be an option to make all Ion requests execute as cold async operations".

I agree it's just silly to make Ion depend on RxAndroid when a conversion is just one line of code away. Then again.. @schwiz's suggestion isn't that bad either.

**Comment by schwiz** [comment]
I'm not sure of the internals but it would have to block before calling
onNext with Observable.from(future)

On Fri, Jun 5, 2015 at 3:21 AM, Frode Nilsen <notifications@github.com>
wrote:

> I'd not recommend using Observable.from(future) since that will block a
> thread somewhere calling future.get.
>
> I'm not sure this is right, or I misunderstand you.
> Observable.from(future) does not block - you're just converting one async
> method to another. Observable.from(future).toBlocking() would on the
> other hand block.
>
> The problem, as I've come to understand, is not with .from() but with the
> fact that Ion is *hot* while Retrofit is *cold* ([hot vs cold])(
> http://www.introtorx.com/content/v1.0.10621.0/14_HotAndColdObservables.html)).
> Well - it's only a problem if you care about interchangability. I'm
> actually inclined to change my issue #395
> <https://github.com/koush/ion/issues/395> to address that matter instead
> - changin the title to something like "There should be an option to make
> all Ion requests execute as cold async operations".
>
> I agree it's just silly to make Ion depend on RxAndroid when a conversion
> is just one line of code away. Then again.. @schwiz
> <https://github.com/schwiz>'s suggestion isn't that bad either.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/koush/ion/issues/545#issuecomment-109200839>.
>



-- 
*Nathan Schwermann*
*785-312-0080*
*http://schwiz.net <http://schwiz.net>*

**Comment by kingsleyadio** [comment]
But one could use:
```java
Observable.defer(() -> 
    Observable.just(ionRequestThatReturnsFuture()))
  .map(response -> 
    { /* some stuff */ })
  .subscribeOn(Schedulers.io())
  .observeOn(AndroidSchedulers.mainThread())
  .subscribe(/*...*/);
```
?

**Comment by schwiz** [comment]
Its best to use Observable.create and have the subscibe method you make
implement the ion callback

**Comment by adiktz** [comment]
Do we have a conclusion?

**Comment by dave08** [comment]
This is the code I use in Kotlin, I think it avoids all these problems...?:

'''
fun <T> ResponseFuture<T>.rxSingle(): Single<T> =
		Single.fromEmitter({ emitter ->
			setCallback { e, result ->
				if (result != null)
					emitter.onSuccess(result)
				else
					emitter.onError(e)
			}

			emitter.setCancellation { cancel() }
		})
'''
Then I can just chain with the request:
'''
		Ion.with(context)
				.load(UrlFormater.getQueryAllCategories(deviceAccountId.toString()))
				.asString()
				.rxSingle()
				.map { GsonProvider.getInstance().parse(it, Category::class.java) as List<Category> }
'''
